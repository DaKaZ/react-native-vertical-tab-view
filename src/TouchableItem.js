/* @flow */

import * as React from 'react';
import {
  TouchableNativeFeedback,
  TouchableOpacity,
  Platform,
  View,
} from 'react-native';
import type { ViewStyleProp } from 'react-native/Libraries/StyleSheet/StyleSheet';

const LOLLIPOP = 21;

type Props = {
  onPress: () => mixed,
  delayPressIn?: number,
  borderless?: boolean,
  pressColor?: string,
  pressOpacity?: number,
  children?: React.Node,
  style?: ViewStyleProp,
};

export default class TouchableItem extends React.Component<Props> {
  static defaultProps = {
    pressColor: 'rgba(255, 255, 255, .4)',
  };

  render() {
    const { style, pressOpacity, pressColor, borderless, ...rest } = this.props;

    if (Platform.OS === 'android' && Platform.Version >= LOLLIPOP) {
      return (
        <TouchableNativeFeedback
          {...rest}
          background={TouchableNativeFeedback.Ripple(pressColor, borderless)}
        >
          <View style={style}>{React.Children.only(this.props.children)}</View>
        </TouchableNativeFeedback>
      );
    } else {
      return (
        <TouchableOpacity {...rest} style={style} activeOpacity={pressOpacity}>
          {this.props.children}
        </TouchableOpacity>
      );
    }
  }
}
