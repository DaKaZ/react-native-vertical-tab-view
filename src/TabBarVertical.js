/* @flow */

import * as React from 'react';
import {
  Animated,
  NativeModules,
  StyleSheet,
  View,
  ScrollView,
  Platform,
  I18nManager,
} from 'react-native';
import TouchableItem from './TouchableItem';
import type { Scene, SceneRendererProps } from './index';
import type {
  ViewStyleProp,
  TextStyleProp,
} from 'react-native/Libraries/StyleSheet/StyleSheet';

type IndicatorProps<T> = SceneRendererProps<T> & {
  width: number,
  height: number,
};

type Props<T> = SceneRendererProps<T> & {
  scrollEnabled?: boolean,
  allowBorder?: boolean,
  bounces?: boolean,
  pressColor?: string,
  pressOpacity?: number,
  getLabelText: (scene: Scene<T>) =>?string,
  getAccessible: (scene: Scene<T>) =>?boolean,
  getAccessibilityLabel: (scene: Scene<T>) =>?string,
  getTestID: (scene: Scene<T>) =>?string,
  renderLabel?: (scene: Scene<T>) => React.Node,
  renderIcon?: (scene: Scene<T>) => React.Node,
  renderBadge?: (scene: Scene<T>) => React.Node,
  c?: (props: IndicatorProps<T>) => React.Node,
  onTabPress?: (scene: Scene<T>) => mixed,
  tabStyle?: ViewStyleProp, // Be sure to include height!
  indicatorStyle?: ViewStyleProp,
  borderStyle?: ViewStyleProp,
  labelStyle?: TextStyleProp,
  style?: ViewStyleProp,
};

type State = {|
  visibility: Animated.Value,
  scrollAmount: Animated.Value,
  initialOffset: ?{| x: number, y: number |},
|};

const useNativeDriver = Boolean(NativeModules.NativeAnimatedModule);

export default class TabBarVertical<T: *> extends React.Component<
  Props<T>,
  State
> {
  static defaultProps = {
    getLabelText: ({ route }: Scene<T>) =>
      typeof route.title === 'string' ? route.title.toUpperCase() : route.title,
    getAccessible: ({ route }: Scene<T>) =>
      typeof route.accessible !== 'undefined' ? route.accessible : true,
    getAccessibilityLabel: ({ route }: Scene<T>) => route.accessibilityLabel,
    getTestID: ({ route }: Scene<T>) => route.testID,
  };

  constructor(props: Props<T>) {
    super(props);

    let initialVisibility = 1;

    if (this.props.scrollEnabled) {
      const tabHeight = this._getTabHeight(this.props);
      if (!tabHeight) {
        initialVisibility = 0;
      }
    }
    const initialOffset =
      props.scrollEnabled && props.layout.width
        ? {
            y: this._getScrollAmount(props, props.navigationState.index),
            x: 0,
          }
        : undefined;

    this.state = {
      visibility: new Animated.Value(initialVisibility),
      scrollAmount: new Animated.Value(0),
      initialOffset,
    };
  }

  componentDidMount() {
    this.props.scrollEnabled && this._startTrackingPosition();
  }

  componentDidUpdate(prevProps: Props<T>) {
    const prevTabWidth = this._getTabWidth(prevProps);
    const currentTabWidth = this._getTabWidth(this.props);
    const pendingIndex =
      typeof this._pendingIndex === 'number'
        ? this._pendingIndex
        : this.props.navigationState.index;

    this._pendingIndex = null;

    if (prevTabWidth !== currentTabWidth && currentTabWidth) {
      this.state.visibility.setValue(1);
    }

    if (
      prevProps.navigationState.routes.length !==
        this.props.navigationState.routes.length ||
      prevProps.layout.width !== this.props.layout.width
    ) {
      this._resetScroll(this.props.navigationState.index, false);
    } else if (prevProps.navigationState.index !== pendingIndex) {
      this._resetScroll(this.props.navigationState.index);
    }
  }

  componentWillUnmount() {
    this._stopTrackingPosition();
  }

  _scrollView: ?ScrollView;
  _isIntial: boolean = true;
  _isManualScroll: boolean = false;
  _isMomentumScroll: boolean = false;
  _pendingIndex: ?number;
  _scrollResetCallback: any;
  _lastPanX: ?number;
  _lastOffsetX: ?number;
  _panXListener: string;
  _offsetXListener: string;

  _startTrackingPosition = () => {
    this._offsetXListener = this.props.offsetX.addListener(({ value }) => {
      this._lastOffsetX = value;
      this._handlePosition();
    });
    this._panXListener = this.props.panX.addListener(({ value }) => {
      this._lastPanX = value;
      this._handlePosition();
    });
  };

  _stopTrackingPosition = () => {
    this.props.offsetX.removeListener(this._offsetXListener);
    this.props.panX.removeListener(this._panXListener);
  };

  _handlePosition = () => {
    const { navigationState, layout } = this.props;

    if (layout.width === 0) {
      // Don't do anything if we don't have layout yet
      return;
    }

    const panX = typeof this._lastPanX === 'number' ? this._lastPanX : 0;
    const offsetX =
      typeof this._lastOffsetX === 'number'
        ? this._lastOffsetX
        : -navigationState.index * layout.width;

    const value = (panX + offsetX) / -(layout.width || 0.001);

    this._adjustScroll(value);
  };

  _renderLabel = (scene: Scene<*>) => {
    if (typeof this.props.renderLabel !== 'undefined') {
      return this.props.renderLabel(scene);
    }
    const label = this.props.getLabelText(scene);
    if (typeof label !== 'string') {
      return null;
    }
    return (
      <Animated.Text style={[styles.tabLabel, this.props.labelStyle]}>
        {label}
      </Animated.Text>
    );
  };

  _renderIndicator = (props: IndicatorProps<T>) => {
    if (typeof this.props.renderIndicator !== 'undefined') {
      return this.props.renderIndicator(props);
    }
    const { width, height, position, navigationState } = props;
    const translateY = Animated.multiply(
      Animated.multiply(
        position.interpolate({
          inputRange: [0, navigationState.routes.length - 1],
          outputRange: [0, navigationState.routes.length - 1],
          extrapolate: 'clamp',
        }),
        height
      ),
      I18nManager.isRTL ? -1 : 1
    );
    return (
      <Animated.View
        style={[
          styles.indicator,
          { width, transform: [{ translateY }] },
          this.props.indicatorStyle,
        ]}
      />
    );
  };

  _getTabWidth = props => {
    const { layout, navigationState, tabStyle } = props;
    const flattened = StyleSheet.flatten(tabStyle);

    if (flattened) {
      switch (typeof flattened.width) {
        case 'number':
          return flattened.width;
        case 'string':
          if (flattened.width.endsWith('%')) {
            const width = parseFloat(flattened.width);
            if (Number.isFinite(width)) {
              return layout.width * (width / 100);
            }
          }
      }
    }

    if (props.scrollEnabled) {
      return (layout.width / 5) * 2;
    }

    return layout.width / navigationState.routes.length;
  };

  _getTabHeight = props => {
    const { layout, navigationState, tabStyle } = props;
    const flattened = StyleSheet.flatten(tabStyle);

    if (flattened) {
      switch (typeof flattened.height) {
        case 'number':
          return flattened.height;
        case 'string':
          if (flattened.height.endsWith('%')) {
            const height = parseFloat(flattened.height);
            if (Number.isFinite(height)) {
              return layout.height * (height / 100);
            }
          }
      }
    }

    if (props.scrollEnabled) {
      return (layout.height / 5) * 2;
    }
    return layout.height / navigationState.routes.length;
  };

  _handleTabPress = ({ route }: Scene<*>) => {
    this._pendingIndex = this.props.navigationState.routes.indexOf(route);

    if (this.props.onTabPress) {
      this.props.onTabPress({ route });
    }

    this.props.jumpTo(route.key);
  };

  _normalizeScrollValue = (props, value) => {
    const { layout, navigationState } = props;
    const tabHeight = this._getTabHeight(props);
    const tabBarHeight = Math.max(
      tabHeight * navigationState.routes.length,
      layout.height
    );
    const maxDistance = tabBarHeight - layout.height;

    return Math.max(Math.min(value, maxDistance), 0);
  };

  _getScrollAmount = (props, i) => {
    const { layout } = props;
    const tabHeight = this._getTabHeight(props);
    const centerDistance = tabHeight * (i + 1 / 2);
    const scrollAmount = centerDistance - layout.height / 2;

    return this._normalizeScrollValue(props, scrollAmount);
  };

  _adjustScroll = (value: number) => {
    if (this.props.scrollEnabled) {
      global.cancelAnimationFrame(this._scrollResetCallback);
      this._scrollView &&
        this._scrollView.scrollTo({
          y: this._normalizeScrollValue(
            this.props,
            this._getScrollAmount(this.props, value)
          ),
          animated: !this._isIntial, // Disable animation for the initial render
        });

      this._isIntial = false;
    }
  };

  _resetScroll = (value: number, animated = true) => {
    if (this.props.scrollEnabled) {
      global.cancelAnimationFrame(this._scrollResetCallback);
      this._scrollResetCallback = global.requestAnimationFrame(() => {
        this._scrollView &&
          this._scrollView.scrollTo({
            y: this._getScrollAmount(this.props, value),
            animated,
          });
      });
    }
  };

  _handleBeginDrag = () => {
    // onScrollBeginDrag fires when user touches the ScrollView
    this._isManualScroll = true;
    this._isMomentumScroll = false;
  };

  _handleEndDrag = () => {
    // onScrollEndDrag fires when user lifts his finger
    // onMomentumScrollBegin fires after touch end
    // run the logic in next frame so we get onMomentumScrollBegin first
    global.requestAnimationFrame(() => {
      if (this._isMomentumScroll) {
        return;
      }
      this._isManualScroll = false;
    });
  };

  _handleMomentumScrollBegin = () => {
    // onMomentumScrollBegin fires on flick, as well as programmatic scroll
    this._isMomentumScroll = true;
  };

  _handleMomentumScrollEnd = () => {
    // onMomentumScrollEnd fires when the scroll finishes
    this._isMomentumScroll = false;
    this._isManualScroll = false;
  };

  render() {
    const { position, navigationState, scrollEnabled, bounces } = this.props;
    const { routes } = navigationState;
    const tabWidth = this._getTabWidth(this.props);

    const tabHeight = this._getTabHeight(this.props);

    // Prepend '-1', so there are always at least 2 items in inputRange
    const inputRange = [-1, ...routes.map((x, i) => i)];
    const translateY = Animated.multiply(this.state.scrollAmount, -1);

    return (
      <Animated.View style={[styles.tabBar, this.props.style]}>
        <Animated.View
          pointerEvents="none"
          style={[
            styles.indicatorContainer,
            scrollEnabled
              ? { height: tabHeight, transform: [{ translateY }] }
              : null,
          ]}
        >
          {this._renderIndicator({
            ...this.props,
            width: tabWidth,
            height: tabHeight,
          })}
        </Animated.View>
        <View style={styles.scroll}>
          <Animated.ScrollView
            horizontal={false}
            keyboardShouldPersistTaps="handled"
            scrollEnabled={scrollEnabled}
            bounces={bounces}
            alwaysBounceHorizontal={false}
            scrollsToTop={false}
            showsHorizontalScrollIndicator={false}
            automaticallyAdjustContentInsets={false}
            overScrollMode="never"
            contentContainerStyle={[
              styles.tabContent,
              scrollEnabled ? null : styles.container,
            ]}
            scrollEventThrottle={1}
            onScroll={Animated.event(
              [
                {
                  nativeEvent: {
                    contentOffset: { y: this.state.scrollAmount },
                  },
                },
              ],
              { useNativeDriver }
            )}
            onScrollBeginDrag={this._handleBeginDrag}
            onScrollEndDrag={this._handleEndDrag}
            onMomentumScrollBegin={this._handleMomentumScrollBegin}
            onMomentumScrollEnd={this._handleMomentumScrollEnd}
            contentOffset={this.state.initialOffset}
            ref={el => (this._scrollView = el && el._component)}
          >
            {routes.map((route, i) => {
              const outputRange = inputRange.map(
                inputIndex => (inputIndex === i ? 1 : 0.7)
              );
              const opacity = Animated.multiply(
                this.state.visibility,
                position.interpolate({
                  inputRange,
                  outputRange,
                })
              );
              const label = this._renderLabel({ route });
              const icon = this.props.renderIcon
                ? this.props.renderIcon({ route })
                : null;
              const badge = this.props.renderBadge
                ? this.props.renderBadge({ route })
                : null;

              const tabStyle = {};

              tabStyle.opacity = opacity;

              if (allowBorder) {

                const borderStyleProps = Object.keys(this.props.borderStyle);
                let borderWidth = 0;
                let borderWidthProp = borderStyleProps.filter((prop) => prop.indexOf('Width') != -1)[0];

                borderWidth = this.props.borderStyle[borderWidthProp];

                outputRange = inputRange.map(
                  inputIndex => (inputIndex === i ? parseInt(borderWidth) : 0)
                );

                borderWidth = Animated.multiply(
                  this.state.visibility,
                  position.interpolate({
                    inputRange,
                    outputRange,
                  })
                );
                
                tabStyle = Object.assign(tabStyle, StyleSheet.flatten(borderStyle))
                tabStyle[borderWidthProp] = borderWidth;
              }

              if (icon) {
                if (label) {
                  tabStyle.paddingTop = 8;
                } else {
                  tabStyle.padding = 12;
                }
              }

              const passedTabStyle = StyleSheet.flatten(this.props.tabStyle);
              const isWidthSet =
                (passedTabStyle &&
                  typeof passedTabStyle.width !== 'undefined') ||
                scrollEnabled === true;
              const tabContainerStyle = {};

              if (isWidthSet) {
                tabStyle.width = tabWidth;
              }

              if (passedTabStyle && typeof passedTabStyle.flex === 'number') {
                tabContainerStyle.flex = passedTabStyle.flex;
              } else if (!isWidthSet) {
                tabContainerStyle.flex = 1;
              }

              let accessibilityLabel = this.props.getAccessibilityLabel({
                route,
              });

              accessibilityLabel =
                typeof accessibilityLabel !== 'undefined'
                  ? accessibilityLabel
                  : this.props.getLabelText({ route });

              const isFocused = i === navigationState.index;

              return (
                <TouchableItem
                  borderless
                  key={route.key}
                  testID={this.props.getTestID({ route })}
                  accessible={this.props.getAccessible({ route })}
                  accessibilityLabel={accessibilityLabel}
                  accessibilityTraits={
                    isFocused ? ['button', 'selected'] : 'button'
                  }
                  accessibilityComponentType="button"
                  pressColor={this.props.pressColor}
                  pressOpacity={this.props.pressOpacity}
                  delayPressIn={0}
                  onPress={() => this._handleTabPress({ route })}
                  style={tabContainerStyle}
                >
                  <View pointerEvents="none" style={styles.container}>
                    <Animated.View
                      style={[
                        styles.tabItem,
                        tabStyle,
                        passedTabStyle,
                        styles.container,
                      ]}
                    >
                      {icon}
                      {label}
                    </Animated.View>
                    {badge ? (
                      <Animated.View
                        style={[
                          styles.badge,
                          { opacity: this.state.visibility },
                        ]}
                      >
                        {badge}
                      </Animated.View>
                    ) : null}
                  </View>
                </TouchableItem>
              );
            })}
          </Animated.ScrollView>
        </View>
      </Animated.View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  scroll: {
    overflow: Platform.OS === 'web' ? ('auto': any) : 'scroll',
  },
  tabBar: {
    backgroundColor: '#2196f3',
    elevation: 4,
    flexDirection: 'row',
    shadowColor: 'black',
    shadowOpacity: 0.1,
    shadowRadius: StyleSheet.hairlineWidth,
    shadowOffset: {
      height: StyleSheet.hairlineWidth,
    },
    // We don't need zIndex on Android, disable it since it's buggy
    zIndex: Platform.OS === 'android' ? 0 : 1,
  },
  tabContent: {
    flexDirection: 'column',
    flexWrap: 'nowrap',
  },
  tabLabel: {
    backgroundColor: 'transparent',
    color: 'white',
    margin: 8,
  },
  tabItem: {
    flex: 1,
    padding: 8,
    alignItems: 'center',
    justifyContent: 'center',
  },
  badge: {
    position: 'absolute',
    top: 0,
    right: 0,
  },
  indicatorContainer: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
  },
  indicator: {
    backgroundColor: '#ffeb3b',
    position: 'absolute',
    left: 0,
    bottom: 0,
    right: 0,
    height: 2,
  },
});
