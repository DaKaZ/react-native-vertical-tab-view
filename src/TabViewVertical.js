/* @flow */

import * as React from 'react';
import { Animated, View, StyleSheet } from 'react-native';
import TabBarVertical from './TabBarVertical';
import { PagerDefault } from 'react-native-tab-view';

import type {
  Scene,
  SceneRendererProps,
  NavigationState,
  Layout,
  PagerCommonProps,
  PagerExtraProps,
} from 'react-native-tab-view/src/TypeDefinitions';
import type { ViewStyleProp } from 'react-native/Libraries/StyleSheet/StyleSheet';

type Props<T> = PagerCommonProps<T> &
  PagerExtraProps & {
    navigationState: NavigationState<T>,
    onIndexChange: (index: number) => mixed,
    initialLayout?: Layout,
    renderPager: (props: *) => React.Node,
    renderScene: (props: SceneRendererProps<T> & Scene<T>) => React.Node,
    renderTabBar: (props: SceneRendererProps<T>) => React.Node,
    tabBarPosition: 'top' | 'bottom',
    useNativeDriver?: boolean,
    style?: ViewStyleProp,
  };

type State = {|
  sceneLayout: Layout & { measured: boolean },
  windowLayout: Layout & { measured: boolean },
  layoutXY: Animated.ValueXY,
  panX: Animated.Value,
  offsetX: Animated.Value,
  position: any,
|};

export default class TabViewVertical<T: *> extends React.Component<
  Props<T>,
  State
> {
  static defaultProps = {
    canJumpToTab: () => true,
    tabBarPosition: 'top',
    renderTabBar: (props: *) => <TabBarVertical {...props} />,
    renderPager: (props: *) => <PagerDefault {...props} />,
    getTestID: ({ route }: Scene<*>) =>
      typeof route.testID === 'string' ? route.testID : undefined,
    initialLayout: {
      height: 0,
      width: 0,
    },
    useNativeDriver: false,
  };

  constructor(props: Props<T>) {
    super(props);

    const { navigationState } = this.props;
    const layout = {
      ...this.props.initialLayout,
      measured: false,
    };

    const panX = new Animated.Value(0);
    const offsetX = new Animated.Value(-navigationState.index * layout.width);
    const layoutXY = new Animated.ValueXY({
      // This is hacky, but we need to make sure that the value is never 0
      x: layout.width || 0.001,
      y: layout.height || 0.001,
    });
    const position = Animated.multiply(
      Animated.divide(Animated.add(panX, offsetX), layoutXY.x),
      -1
    );

    this.state = {
      sceneLayout: layout,
      windowLayout: layout,
      layoutXY,
      panX,
      offsetX,
      position,
    };
  }

  componentDidMount() {
    this._mounted = true;
  }

  componentWillUnmount() {
    this._mounted = false;
  }

  _mounted: boolean = false;
  _nextIndex: ?number;

  _renderScene = (props: SceneRendererProps<T> & Scene<T>) => {
    return this.props.renderScene(props);
  };

  _handleLayoutScene = (e: any) => {
    const { height, width } = e.nativeEvent.layout;

    if (
      this.state.sceneLayout.width === width &&
      this.state.sceneLayout.height === height
    ) {
      return;
    }

    this.state.offsetX.setValue(-this.props.navigationState.index * width);
    this.state.layoutXY.setValue({
      // This is hacky, but we need to make sure that the value is never 0
      x: width || 0.001,
      y: height || 0.001,
    });

    this.setState({
      sceneLayout: {
        measured: true,
        height,
        width,
      },
    });
  };

  _handleLayoutWindow = (e: any) => {
    const { height, width } = e.nativeEvent.layout;

    if (
      this.state.windowLayout.width === width &&
      this.state.windowLayout.height === height
    ) {
      return;
    }

    this.setState({
      windowLayout: {
        measured: true,
        height,
        width,
      },
    });
  };

  _buildSceneRendererProps = (): SceneRendererProps<*> => ({
    panX: this.state.panX,
    offsetX: this.state.offsetX,
    position: this.state.position,
    layout: this.state.sceneLayout,
    navigationState: this.props.navigationState,
    jumpTo: this._jumpTo,
    useNativeDriver: this.props.useNativeDriver === true,
  });

  _jumpTo = (key: string) => {
    if (!this._mounted) {
      // We are no longer mounted, this is a no-op
      return;
    }

    const { canJumpToTab, navigationState } = this.props;
    const index = navigationState.routes.findIndex(route => route.key === key);

    if (!canJumpToTab(navigationState.routes[index])) {
      return;
    }

    if (index !== navigationState.index) {
      this.props.onIndexChange(index);
    }
  };

  render() {
    const {
      /* eslint-disable no-unused-vars */
      navigationState,
      onIndexChange,
      initialLayout,
      renderScene,
      /* eslint-enable no-unused-vars */
      renderPager,
      renderTabBar,
      tabBarPosition,
      ...rest
    } = this.props;

    const props = this._buildSceneRendererProps();

    return (
      <View
        collapsable={false}
        onLayout={this._handleLayoutWindow}
        style={[styles.container, this.props.style]}
      >
        {(tabBarPosition === 'top' || tabBarPosition === 'left') &&
          renderTabBar(props)}
        <View onLayout={this._handleLayoutScene} style={styles.pager}>
          {renderPager({
            ...props,
            ...rest,
            panX: this.state.panX,
            offsetX: this.state.offsetX,
            children: navigationState.routes.map(route => {
              const scene = this._renderScene({
                ...props,
                route,
              });

              if (React.isValidElement(scene)) {
                /* $FlowFixMe: https://github.com/facebook/flow/issues/4775 */
                return React.cloneElement(scene, { key: route.key });
              }

              return scene;
            }),
          })}
        </View>
        {(tabBarPosition === 'bottom' || tabBarPosition === 'right') &&
          renderTabBar(props)}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    overflow: 'hidden',
    flexDirection: 'row',
  },
  pager: {
    flex: 1,
  },
});
