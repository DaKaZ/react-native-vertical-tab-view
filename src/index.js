/* @flow */

import {
  TabView,
  TabBar,
  PagerDefault,
  PagerScroll,
  PagerAndroid,
  PagerExperimental,
  SceneMap,
} from 'react-native-tab-view';

import TabViewVertical from './TabViewVertical';
import TabBarVertical from './TabBarVertical';

export {
  TabView,
  TabViewVertical,
  TabBar,
  TabBarVertical,
  PagerDefault,
  PagerScroll,
  PagerAndroid,
  PagerExperimental,
  SceneMap,
};

export type {
  Route,
  Scene,
  NavigationState,
  SceneRendererProps,
} from 'react-native-tab-view';
