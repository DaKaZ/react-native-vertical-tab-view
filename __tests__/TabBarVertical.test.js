/* @flow */

import * as React from 'react';
import { StyleSheet, View } from 'react-native';
import { shallow } from 'enzyme';
import TabViewVertical from '../src/TabViewVertical';
import TabBarVertical from '../src/TabBarVertical';
import { SceneMap } from '../src/index';

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  tabbar: {
    backgroundColor: '#3f51b5',
  },
  tab: {
    width: 180,
    height: 80,
  },
  indicator: {
    backgroundColor: '#ffeb3b',
  },
  label: {
    color: '#fff',
    fontWeight: '400',
  },
});

const state = {
  index: 1,
  routes: [
    { key: 'article', title: 'Article' },
    { key: 'contacts', title: 'Contacts' },
    { key: 'albums', title: 'Albums' },
    { key: 'chat', title: 'Chat' },
  ],
};

const _handleIndexChange = jest.fn();

const _renderTabBar = props => (
  <TabBarVertical
    {...props}
    scrollEnabled
    indicatorStyle={styles.indicator}
    style={styles.tabbar}
    tabStyle={styles.tab}
    labelStyle={styles.label}
  />
);

const _renderScene = SceneMap({
  albums: View,
  contacts: View,
  article: View,
  chat: View,
});

it('renders a vertical tab bar', () => {
  const component = shallow(
    <TabViewVertical
      style={[styles.container]}
      navigationState={state}
      renderScene={_renderScene}
      renderTabBar={_renderTabBar}
      onIndexChange={_handleIndexChange}
      initialLayout={{ width: 800, height: 600 }}
    />
  );
  expect(component).toMatchSnapshot();
});
