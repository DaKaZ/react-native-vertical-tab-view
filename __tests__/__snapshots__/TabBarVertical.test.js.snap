// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a vertical tab bar 1`] = `
<Component
  collapsable={false}
  onLayout={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
        "flexDirection": "row",
        "overflow": "hidden",
      },
      Array [
        Object {
          "flex": 1,
        },
      ],
    ]
  }
>
  <TabBarVertical
    getAccessibilityLabel={[Function]}
    getAccessible={[Function]}
    getLabelText={[Function]}
    getTestID={[Function]}
    indicatorStyle={
      Object {
        "backgroundColor": "#ffeb3b",
      }
    }
    jumpTo={[Function]}
    labelStyle={
      Object {
        "color": "#fff",
        "fontWeight": "400",
      }
    }
    layout={
      Object {
        "height": 600,
        "measured": false,
        "width": 0,
      }
    }
    navigationState={
      Object {
        "index": 1,
        "routes": Array [
          Object {
            "key": "article",
            "title": "Article",
          },
          Object {
            "key": "contacts",
            "title": "Contacts",
          },
          Object {
            "key": "albums",
            "title": "Albums",
          },
          Object {
            "key": "chat",
            "title": "Chat",
          },
        ],
      }
    }
    offsetX={-800}
    panX={0}
    position={1}
    scrollEnabled={true}
    style={
      Object {
        "backgroundColor": "#3f51b5",
      }
    }
    tabStyle={
      Object {
        "height": 80,
        "width": 180,
      }
    }
    useNativeDriver={false}
  />
  <Component
    onLayout={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <PagerScroll
      canJumpToTab={[Function]}
      getTestID={[Function]}
      jumpTo={[Function]}
      layout={
        Object {
          "height": 600,
          "measured": false,
          "width": 800,
        }
      }
      navigationState={
        Object {
          "index": 1,
          "routes": Array [
            Object {
              "key": "article",
              "title": "Article",
            },
            Object {
              "key": "contacts",
              "title": "Contacts",
            },
            Object {
              "key": "albums",
              "title": "Albums",
            },
            Object {
              "key": "chat",
              "title": "Chat",
            },
          ],
        }
      }
      offsetX={-800}
      panX={0}
      position={1}
      style={
        Array [
          Object {
            "flex": 1,
          },
        ]
      }
      useNativeDriver={false}
    >
      <SceneComponent
        component={[Function]}
        jumpTo={[Function]}
        key="article"
        route={
          Object {
            "key": "article",
            "title": "Article",
          }
        }
      />
      <SceneComponent
        component={[Function]}
        jumpTo={[Function]}
        key="contacts"
        route={
          Object {
            "key": "contacts",
            "title": "Contacts",
          }
        }
      />
      <SceneComponent
        component={[Function]}
        jumpTo={[Function]}
        key="albums"
        route={
          Object {
            "key": "albums",
            "title": "Albums",
          }
        }
      />
      <SceneComponent
        component={[Function]}
        jumpTo={[Function]}
        key="chat"
        route={
          Object {
            "key": "chat",
            "title": "Chat",
          }
        }
      />
    </PagerScroll>
  </Component>
</Component>
`;
